eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -286,12 +287,11 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo)
     {
         $dbHandler = $this->getConnection();
         // do not delete assignments if there's other published version containing this field
-        if ($versionInfo !== null &&
-            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+        if ($versionInfo->status === VersionInfo::STATUS_ARCHIVED &&
```

Comment:
I just tried to do it like that, but still no success, other tests fail including basic ones: `testAddFieldDefinition`, `testUpdatedDataCorrect`.
It's worth mentioning that per one content create/update `deleteOldKeywordAssignments` method is called several (~3) times, meaning that maybe applying proper logic here is not enough and something needs to be refactored on a higher level.

Also the test case with deleting draft of published version fails without my changes, so this is yet another bug...